File: internal/ml/pattern_matcher.go
package ml

import (
	"encoding/json"
	"fmt"
	"math"
	"os"
	"sort"
)

// Patterns mirrors pattern_matcher.py PatternIndex serialization.
// Scaler stored as mean/scale arrays; samples pre-scaled.
// Linear k-NN scan (n typically 5-10k, k=10 → fast enough).
type Patterns struct {
	FeatureCols []string    `json:"feature_cols"`
	ScalerMean  []float64   `json:"scaler_mean"`
	ScalerScale []float64   `json:"scaler_scale"`
	Samples     [][]float64 `json:"samples"`   // N × F, already z-scored
	Outcomes    [][]float64 `json:"outcomes"`  // N × 3 (5/10/20 bar returns)
	Timestamps  []string    `json:"timestamps"`
	Closes      []float64   `json:"closes"`
}

type SimilarSituation struct {
	Date        string  `json:"date"`
	Distance    float64 `json:"distance"`
	Outcome5    float64 `json:"outcome_5"`
	Outcome10   float64 `json:"outcome_10"`
	Outcome20   float64 `json:"outcome_20"`
	Description string  `json:"description"`
}

func readPatternsFile(path string) (*Patterns, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var p Patterns
	if err := json.Unmarshal(data, &p); err != nil {
		return nil, err
	}
	return &p, nil
}

// Query returns top-k most similar historical samples.
// rawFeatures are in original feature space — we z-score them with stored scaler.
func (p *Patterns) Query(rawFeatures []float64, k int) []SimilarSituation {
	// Strict validation to prevent index-out-of-bounds panic when
	// model and scaler arrays have mismatched lengths (corrupted JSON).
	if len(p.Samples) == 0 ||
		len(rawFeatures) != len(p.FeatureCols) ||
		len(p.ScalerMean) != len(rawFeatures) ||
		len(p.ScalerScale) != len(rawFeatures) {
		return nil
	}
	// Scale.
	x := make([]float64, len(rawFeatures))
	for i, v := range rawFeatures {
		scale := p.ScalerScale[i]
		if scale == 0 {
			scale = 1
		}
		x[i] = (v - p.ScalerMean[i]) / scale
	}

	// Compute distances (linear; fine for <50k samples).
	type hit struct {
		idx  int
		dist float64
	}
	hits := make([]hit, len(p.Samples))
	for i, s := range p.Samples {
		var d float64
		for j, v := range s {
			diff := v - x[j]
			d += diff * diff
		}
		hits[i] = hit{idx: i, dist: math.Sqrt(d)}
	}
	sort.Slice(hits, func(i, j int) bool { return hits[i].dist < hits[j].dist })

	if k > len(hits) {
		k = len(hits)
	}
	out := make([]SimilarSituation, 0, k)
	for i := 0; i < k; i++ {
		h := hits[i]
		out5 := p.Outcomes[h.idx][0]
		out10 := p.Outcomes[h.idx][1]
		out20 := p.Outcomes[h.idx][2]
		out = append(out, SimilarSituation{
			Date:        p.Timestamps[h.idx],
			Distance:    h.dist,
			Outcome5:    out5,
			Outcome10:   out10,
			Outcome20:   out20,
			Description: describeOutcome(h.dist, out10),
		})
	}
	return out
}

func describeOutcome(distance, outcome10 float64) string {
	var closeness string
	switch {
	case distance < 1:
		closeness = "очень похожая"
	case distance < 2:
		closeness = "схожая"
	default:
		closeness = "отдалённая"
	}
	sign := "+"
	pct := outcome10 * 100
	if pct < 0 {
		sign = ""
	}
	return fmt.Sprintf("%s ситуация → %s%.1f%% за 10 баров", closeness, sign, pct)
}

// AggregateOutcome averages the 5/10/20-bar returns across neighbours,
// weighted by inverse distance. Used for the "average outcome" summary in UI.
func AggregateOutcome(sims []SimilarSituation) (avg5, avg10, avg20 float64) {
	if len(sims) == 0 {
		return 0, 0, 0
	}
	var wsum, s5, s10, s20 float64
	for _, s := range sims {
		w := 1.0 / (1.0 + s.Distance)
		wsum += w
		s5 += s.Outcome5 * w
		s10 += s.Outcome10 * w
		s20 += s.Outcome20 * w
	}
	if wsum == 0 {
		return 0, 0, 0
	}
	return s5 / wsum, s10 / wsum, s20 / wsum
}

File: ml-training/pattern_matcher.py
"""
pattern_matcher.py — k-NN on feature vectors for "similar historical situation".

Stores scaled features + outcomes in a BallTree. At inference, query current
feature vector → top-k nearest historical rows with their realized 5/10/20-day
returns. Used for UI wow-factor: "похожая ситуация была 15 марта → +23%".

The index is also serializable to JSON for Go consumption (flat arrays).
"""
from __future__ import annotations

import json
from dataclasses import dataclass, asdict
from datetime import datetime
from pathlib import Path
from typing import List

import numpy as np
import pandas as pd
from sklearn.neighbors import BallTree
from sklearn.preprocessing import StandardScaler


@dataclass
class SimilarSituation:
    date: str            # ISO-8601
    distance: float      # in scaled feature space
    outcome_5: float     # return 5 bars later
    outcome_10: float    # return 10 bars later
    outcome_20: float    # return 20 bars later
    description: str     # human-readable hint


class PatternIndex:
    """BallTree over z-scored features. Keeps raw outcomes for recall."""

    def __init__(self, feature_cols: List[str]):
        self.feature_cols = feature_cols
        self.scaler = StandardScaler()
        self.tree: BallTree | None = None
        self.outcomes: np.ndarray | None = None  # shape (N, 3) for 5/10/20
        self.timestamps: List[str] = []
        self.closes: np.ndarray | None = None

    def fit(self, features_df: pd.DataFrame, closes: np.ndarray, timestamps: List[str]) -> "PatternIndex":
        """
        features_df: N rows × len(feature_cols) columns.
        closes:      N-length array of close prices, aligned.
        timestamps:  N ISO-8601 strings.
        """
        X = features_df[self.feature_cols].to_numpy()
        X_scaled = self.scaler.fit_transform(X)

        # Compute forward outcomes (NaN for last 20 bars).
        n = len(closes)
        out = np.full((n, 3), np.nan)
        for horizon_idx, h in enumerate((5, 10, 20)):
            shift = np.roll(closes, -h)
            ret = (shift - closes) / (closes + 1e-12)
            ret[-h:] = np.nan
            out[:, horizon_idx] = ret

        # Only keep rows with complete outcomes.
        mask = ~np.isnan(out).any(axis=1)
        X_scaled = X_scaled[mask]
        self.outcomes = out[mask]
        self.closes = closes[mask]
        self.timestamps = [timestamps[i] for i in np.where(mask)[0]]

        self.tree = BallTree(X_scaled, leaf_size=40)
        return self

    def query(self, current_feats: np.ndarray, k: int = 10) -> List[SimilarSituation]:
        if self.tree is None:
            return []
        x = self.scaler.transform(current_feats.reshape(1, -1))
        distances, indices = self.tree.query(x, k=min(k, len(self.timestamps)))
        results: List[SimilarSituation] = []
        for d, i in zip(distances[0], indices[0]):
            out5, out10, out20 = self.outcomes[i]
            results.append(
                SimilarSituation(
                    date=self.timestamps[i],
                    distance=float(d),
                    outcome_5=float(out5),
                    outcome_10=float(out10),
                    outcome_20=float(out20),
                    description=_describe(d, out10),
                )
            )
        return results

    def to_json_payload(self) -> dict:
        """Flat serialization: Go reads scaler + samples (no BallTree — Go uses linear scan for k≤20)."""
        if self.tree is None:
            return {}
        # Recover scaled features for export.
        # BallTree stores data internally; grab from the tree.
        scaled = np.asarray(self.tree.data)
        return {
            "feature_cols": self.feature_cols,
            "scaler_mean": self.scaler.mean_.tolist(),
            "scaler_scale": self.scaler.scale_.tolist(),
            "samples": scaled.tolist(),  # N × F scaled vectors
            "outcomes": self.outcomes.tolist(),
            "timestamps": self.timestamps,
            "closes": self.closes.tolist() if self.closes is not None else [],
        }

    @classmethod
    def from_json_payload(cls, payload: dict) -> "PatternIndex":
        idx = cls(payload["feature_cols"])
        idx.scaler = StandardScaler()
        idx.scaler.mean_ = np.array(payload["scaler_mean"])
        idx.scaler.scale_ = np.array(payload["scaler_scale"])
        samples = np.array(payload["samples"])
        idx.tree = BallTree(samples, leaf_size=40)
        idx.outcomes = np.array(payload["outcomes"])
        idx.timestamps = payload["timestamps"]
        idx.closes = np.array(payload.get("closes", []))
        return idx


def _describe(distance: float, outcome_10: float) -> str:
    """Generate human-readable hint for UI."""
    closeness = "очень похожая" if distance < 1 else "схожая" if distance < 2 else "отдалённая"
    if outcome_10 > 0.05:
        move = f"+{outcome_10 * 100:.1f}%"
    elif outcome_10 < -0.05:
        move = f"{outcome_10 * 100:.1f}%"
    else:
        move = f"{outcome_10 * 100:+.1f}%"
    return f"{closeness} ситуация → {move} за 10 баров"


def save_index(idx: PatternIndex, path: Path) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    with open(path, "w") as f:
        json.dump(idx.to_json_payload(), f)


def load_index(path: Path) -> PatternIndex:
    with open(path) as f:
        payload = json.load(f)
    return PatternIndex.from_json_payload(payload)
